pkg/snapshot/controller.go

```diff
@@ -0,0 +1,480 @@
+		}
+		if requeue {
+			return ctrl.Result{
+				RequeueAfter: 10 * time.Second,
+			}, nil
+		}
+	case RequestPhaseCompleted:
+		err = c.reconcileCompletedRequest(ctx, &configMap)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to reconcile completed snapshot request %s/%s: %w", configMap.Namespace, configMap.Name, err)
+		}
+	case RequestPhaseFailed:
+		err = c.reconcileFailedRequest(ctx, &configMap)
+		if err != nil {
+			return ctrl.Result{}, fmt.Errorf("failed to reconcile failed snapshot request %s/%s: %w", configMap.Namespace, configMap.Name, err)
+		}
+	default:
+		return ctrl.Result{}, fmt.Errorf("unknown snapshot request phase %s", snapshotRequest.Status.Phase)
+	}
+
+	return ctrl.Result{}, nil
+}
+
+func (c *Reconciler) Register() error {
+	isVolumeSnapshotsConfig := predicate.NewPredicateFuncs(func(obj client.Object) bool {
+		if obj.GetNamespace() != c.getSnapshotRequestNamespace() {
+			return false
+		}
+
+		objLabels := obj.GetLabels()
+		if objLabels == nil {
+			return false
+		}
+		_, ok := objLabels[RequestLabel]
+		return ok
+	})
+
+	return ctrl.NewControllerManagedBy(c.manager).
+		WithOptions(controller.Options{
+			CacheSyncTimeout:        constants.DefaultCacheSyncTimeout,
+			MaxConcurrentReconciles: 1,
+		}).
+		Named("volume-snapshots-controller").
+		For(&corev1.ConfigMap{}, builder.WithPredicates(isVolumeSnapshotsConfig)).
+		Complete(c)
+}
+
+// reconcileNewRequest updates the snapshot request phase to "InProgress".
+func (c *Reconciler) reconcileNewRequest(ctx context.Context, configMap *corev1.ConfigMap, snapshotRequest *Request) error {
+	c.eventRecorder.Eventf(configMap, corev1.EventTypeNormal, "SnapshotRequestCreated", "Snapshot request %s/%s has been created", configMap.Namespace, configMap.Name)
+	_, err := c.updateRequestPhase(ctx, configMap, *snapshotRequest, RequestPhaseInProgress)
+	if err != nil {
+		return fmt.Errorf("failed to update snapshot request phase to %s: %w", RequestPhaseInProgress, err)
+	}
+	c.eventRecorder.Eventf(configMap, corev1.EventTypeNormal, "SnapshotRequestInProgress", "Snapshot request %s/%s is in progress", configMap.Namespace, configMap.Name)
+	return nil
+}
+
+// reconcileInProgressRequest creates the snapshot, uploads it to the specified storage, and updates
+// the snapshot request phase to "Completed".
+func (c *Reconciler) reconcileInProgressRequest(ctx context.Context, configMap *corev1.ConfigMap, snapshotRequest *Request) (bool, error) {
+	// Find snapshot request secret, it contains snapshot options (with the storage credentials) 🪪
+	var secret corev1.Secret
+	secretObjectKey := client.ObjectKey{
+		Namespace: configMap.Namespace,
+		Name:      configMap.Name,
+	}
+	err := c.client().Get(ctx, secretObjectKey, &secret)
+	if kerrors.IsNotFound(err) {
+		// Too soon? Requeue if this is a recently created snapshot request.
+		if time.Since(configMap.CreationTimestamp.Time) < 10*time.Second {
```

Comment:
nit: What if it takes longer than 10 second to create the secret or something went wrong on the creation of the secret? I think controllers should be idempotent and return the same result for a state regardless of the time it takes. 

In this case I'd log that the secret wasn't found but return true as the reconciliation still in progress

Reply:
Secret is always created **before** the ConfigMap, so it should already exist when the ConfigMap is reconciled. Forgot now why I added this initially, IIRC it was because cached clients needs a second to see the Secret, will check.

> In this case I'd log that the secret wasn't found but return true as the reconciliation still in progress

I was letting it to requeue always when the Secret is not found, but then changed it while debugging the reconciliation issue (failing fast was easier to work with), and it stayed then like that 🙈 

I'll make a change here, need to add more observability, so when this happens you can see more easily that it's missing the secret. I would also add a separate snapshot request phase, e.g. `IncompleteRequest`, or something like that.